src/main/java/com/comphenix/protocol/injector/netty/channel/InboundPacketInterceptor.java

```diff
@@ -1,44 +1,52 @@
 package com.comphenix.protocol.injector.netty.channel;
 
+import com.comphenix.protocol.PacketType;
 import com.comphenix.protocol.injector.netty.ChannelListener;
+import com.comphenix.protocol.injector.packet.PacketRegistry;
 import com.comphenix.protocol.utility.MinecraftReflection;
+
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandlerAdapter;
 
 final class InboundPacketInterceptor extends ChannelInboundHandlerAdapter {
 
-    private final NettyChannelInjector injector;
-    private final ChannelListener channelListener;
-
-    public InboundPacketInterceptor(NettyChannelInjector injector, ChannelListener listener) {
-        this.injector = injector;
-        this.channelListener = listener;
-    }
-
-    @Override
-    public void channelRead(ChannelHandlerContext ctx, Object msg) {
-        Class<?> messageClass = msg.getClass();
-        if (this.shouldInterceptMessage(messageClass)) {
-            // process the login if the packet is one before posting the packet to any handler to provide "real" data
-            // the method invocation will do nothing if the packet is not a login packet
-            this.injector.tryProcessLogin(msg);
-
-            // check if there are any listeners bound for the packet - if not just post the packet down the pipeline
-            if (!this.channelListener.hasListener(messageClass)) {
-                ctx.fireChannelRead(msg);
-                return;
-            }
-
-            // call all inbound listeners
-            this.injector.processInboundPacket(ctx, msg, messageClass);
-        } else {
-            // just pass the message down the pipeline
-            ctx.fireChannelRead(msg);
-        }
-    }
-
-    private boolean shouldInterceptMessage(Class<?> messageClass) {
-        // only intercept minecraft packets and no garbage from other stuff in the channel
-        return MinecraftReflection.getPacketClass().isAssignableFrom(messageClass);
-    }
+	private final NettyChannelInjector injector;
+	private final ChannelListener channelListener;
+
+	public InboundPacketInterceptor(NettyChannelInjector injector, ChannelListener listener) {
+		this.injector = injector;
+		this.channelListener = listener;
+	}
+
+	@Override
+	public void channelRead(ChannelHandlerContext ctx, Object msg) {
+		if (MinecraftReflection.isPacketClass(msg)) {
+			// process the login if the packet is one before posting the packet to any
+			// handler to provide "real" data the method invocation will do nothing if the
+			// packet is not a login packet
+			this.injector.tryProcessLogin(msg);
+
+			PacketType.Protocol protocol = this.injector.getInboundProtocol();
+			PacketType packetType = PacketRegistry.getPacketType(protocol, msg.getClass());
+			
+			// TODO: ignore packet or throw error?
```

Comment:
How should unknown packet types get handled? In the old code they would throw an exception but after the PR the would should be silently ignored.

Reply:
silently ignoring in the injector is good. we'd likely only want to throw an exception if a plugin specifically requests an invalid packet type (e.g. in registering their listener)

Reply:
Since I'm checking for null here, should I revert #2961 because this takes care of it (same for https://github.com/Imprex-Development/ProtocolLib/blob/b0c4b7fe45024e59505ed132d2df9d61fd5f405d/src/main/java/com/comphenix/protocol/injector/netty/channel/NettyChannelInjector.java#L489)?

Reply:
assuming it works right on lower versions, then yes

Reply:
Probably would work but I will leave it for now.